bundles/org.openhab.binding.gardena/src/main/java/org/openhab/binding/gardena/internal/GardenaSmartImpl.java

```diff
@@ -12,562 +12,384 @@
+    }
 
-                // determine deleted devices
-                Collection<Device> deletedDevices = allDevicesById.values().stream()
-                        .filter(d -> !newDevicesById.values().contains(d)).collect(Collectors.toSet());
+    /**
+     * Returns the websocket url for a given location.
+     */
+    private WebSocketCreatedResponse getWebsocketInfo(String locationId) throws GardenaException {
+        return executeRequest(HttpMethod.POST, URL_API_WEBSOCKET, new CreateWebSocketRequest(locationId),
+                WebSocketCreatedResponse.class);
+    }
 
-                // determine new devices
-                Collection<Device> newDevices = newDevicesById.values().stream()
-                        .filter(d -> !allDevicesById.values().contains(d)).collect(Collectors.toSet());
+    /**
+     * Stops the client.
+     */
+    public void dispose() {
+        logger.debug("Disposing GardenaSmart");
+        if (restartScheduledFuture != null) {
+            restartScheduledFuture.cancel(true);
+        }
+        if (deviceToNotifyFuture != null) {
+            deviceToNotifyFuture.cancel(true);
+        }
+        stopWebsockets();
+        if (httpClient != null) {
+            try {
+                httpClient.stop();
+            } catch (Exception e) {
+                // ignore
+            }
+            httpClient.destroy();
+        }
+        locationsResponse = new LocationsResponse();
+        allDevicesById.clear();
+        initialized = false;
+    }
 
-                // determine updated devices
-                Collection<Device> updatedDevices = allDevicesById.values().stream().distinct()
-                        .filter(newDevicesById.values()::contains).collect(Collectors.toSet());
+    /**
+     * Restarts all websockets.
+     */
+    private synchronized void restartWebsockets() {
+        logger.debug("Restarting GardenaSmart Webservice");
+        stopWebsockets();
+        try {
+            startWebsockets();
+            eventListener.onConnectionResumed();
+        } catch (Exception ex) {
+            eventListener.onConnectionLost();
+            logger.warn("Restarting GardenaSmart Webservice failed: {}, try restart in {} seconds", ex.getMessage(),
+                    RESTART_DELAY_SECONDS);
+            restartScheduledFuture = scheduler.schedule(() -> {
+                restartWebsockets();
+            }, RESTART_DELAY_SECONDS, TimeUnit.SECONDS);
+        }
+    }
 
-                allDevicesById = newDevicesById;
+    /**
+     * Sets the dataItem from the websocket event into the correct device.
+     */
+    private void handleDataItem(final DataItem dataItem) throws GardenaException {
+        final String deviceId = dataItem.getDeviceId();
+        Device device = allDevicesById.get(deviceId);
+        if (device == null && !(dataItem instanceof LocationDataItem)) {
+            device = new Device(deviceId);
+            allDevicesById.put(device.id, device);
```

Comment:
No, the device id is the id before the colon. So b:1 and b:2 will be b.

Reply:
Such parsing is an attempt to reverse engineer the protocol and might and quite likely will cause problems.
I can't disclose details, but we already had a problem with one client who caused problems to his users by doing exactly this: parsing the IDs.
IDs are not supposed to be parsed, they are opaque tokens.

Reply:
Another anecdotal evidence:
I had to fix another bug today which stemmed exactly from the attempt to assume how different IDs relate to each other and that they match in certain cases. In 99% of cases they behave exactly how you think they do, but then your users find a corner case that you didn't think about.

I see now that it was a mistake to make these IDs parseable because it provides too much of temptation to parse them, but it's too late to change this now, the API is live.